### PR TITLE
OpenZFS 9164 - assert: newds == os->os_dsl_dataset

### PR DESCRIPTION
**Pull request by dinatale2**
Authored by: Andriy Gapon <avg@FreeBSD.org>
Reviewed by: Matt Ahrens <mahrens@delphix.com>
Reviewed by: Don Brady <don.brady@delphix.com>
Approved by: Richard Lowe <richlowe@richlowe.net>
Ported-by: Giuseppe Di Natale <dinatale2@llnl.gov>

OpenZFS-issue: https://www.illumos.org/issues/9164
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/0e776dc06a

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by loli10K** [comment]
This should fix #6112.

**Comment by loli10K** [comment]
"-" is not a valid number and we can't assign it to an integer (`typeset -i`):
```
root@linux:~# zpool list
NAME       SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
testpool   112M   102K   112M         -     0%     0%  1.00x  ONLINE  -
root@linux:~# zpool get version
NAME      PROPERTY  VALUE    SOURCE
testpool  version   -        default
root@linux:~# ksh
# typeset -i vp=0
# vp=$(zpool get version -H -p -o value testpool)
ksh: -: more tokens expected
# str=$(zpool get version -H -p -o value testpool)
# 
root@linux:~# 
```
Tested on ksh version "93u+20120801-1" (Debian8).

Somehow this doesn't seem to affect Illumos: i am going to try and do some more testing on my OpenIndiana development box later today.

**Comment by loli10K** [comment]
> Somehow this doesn't seem to affect Illumos

This **does** affect Illumos: my test box has "/" on ZFS

https://github.com/zfsonlinux/zfs/blob/668173b5764386b47de0beb7c57ff8ce8f68f983/tests/zfs-tests/tests/functional/cli_root/zfs_upgrade/zfs_upgrade.kshlib#L129-L130

in which case the following block of code gets skipped (`vp=$(get_pool_prop version $pool)`)

https://github.com/zfsonlinux/zfs/blob/668173b5764386b47de0beb7c57ff8ce8f68f983/tests/zfs-tests/tests/functional/cli_root/zfs_upgrade/zfs_upgrade.kshlib#L151-L163

**Comment by dinatale2** [comment]
@loli10K That explains it... I wasn't concerned about the assignment. I clearly should have been.

It might look messy, but I could use a temp variable and check for `-`.   That means `default` and I think the default version is currently `5`. Am I wrong?

**Comment by loli10K** [comment]
@dinatale2 i think that should work. Also we should probably wrap those "log_must (zfs|zpool) upgrade >/dev/null" with `eval` to avoid losing `log_must` output.

File: tests/zfs-tests/tests/functional/cli_root/zfs_upgrade/zfs_upgrade.kshlib [hunk 1]
```diff
@@ -42,11 +42,22 @@ set -A spa_create_versions 9 15 24
 set -A zpl_upgrade_versions 3 4  5
 set -A spa_upgrade_versions 9 15 24
 
+function get_pool_version #pool
+{
+	ver=$(get_pool_prop version $1)
+
+	if [ "$ver" = "-" ]; then
+		ver="5000"
```
Comment (loli10K) [inline tests/zfs-tests/tests/functional/cli_root/zfs_upgrade/zfs_upgrade.kshlib]: I'm sorry for derailing this, i was wrongfully thinking of filesystem version, not pool version.

Comment (dinatale2) [reply tests/zfs-tests/tests/functional/cli_root/zfs_upgrade/zfs_upgrade.kshlib]: @loli10K You didn't derail anything. Thank you for your assistance! :)

**Pull request closed without merging**